src/librustc_mir/interpret/memory.rs

```diff
@@ -927,59 +930,58 @@ pub trait HasMemory<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
     fn into_ptr(
         &self,
         value: Value,
-    ) -> EvalResult<'tcx, Pointer> {
+    ) -> EvalResult<'tcx, Scalar> {
         Ok(match value {
             Value::ByRef(ptr, align) => {
                 self.memory().read_ptr_sized(ptr.to_ptr()?, align)?
             }
-            Value::ByVal(ptr) |
-            Value::ByValPair(ptr, _) => ptr,
+            Value::Scalar(ptr) |
+            Value::ScalarPair(ptr, _) => ptr,
         }.into())
     }
 
     fn into_ptr_vtable_pair(
         &self,
         value: Value,
-    ) -> EvalResult<'tcx, (Pointer, MemoryPointer)> {
+    ) -> EvalResult<'tcx, (Scalar, Pointer)> {
         match value {
             Value::ByRef(ref_ptr, align) => {
                 let mem = self.memory();
                 let ptr = mem.read_ptr_sized(ref_ptr.to_ptr()?, align)?.into();
                 let vtable = mem.read_ptr_sized(
-                    ref_ptr.offset(mem.pointer_size(), &mem.tcx.data_layout)?.to_ptr()?,
+                    ref_ptr.ptr_offset(mem.pointer_size(), &mem.tcx.data_layout)?.to_ptr()?,
                     align
                 )?.to_ptr()?;
                 Ok((ptr, vtable))
             }
 
-            Value::ByValPair(ptr, vtable) => Ok((ptr.into(), vtable.to_ptr()?)),
+            Value::ScalarPair(ptr, vtable) => Ok((ptr.into(), vtable.to_ptr()?)),
 
-            Value::ByVal(PrimVal::Undef) => err!(ReadUndefBytes),
+            Value::Scalar(Scalar::Bits { defined: 0, .. }) => err!(ReadUndefBytes),
             _ => bug!("expected ptr and vtable, got {:?}", value),
         }
     }
 
     fn into_slice(
         &self,
         value: Value,
-    ) -> EvalResult<'tcx, (Pointer, u64)> {
+    ) -> EvalResult<'tcx, (Scalar, u64)> {
         match value {
             Value::ByRef(ref_ptr, align) => {
                 let mem = self.memory();
                 let ptr = mem.read_ptr_sized(ref_ptr.to_ptr()?, align)?.into();
                 let len = mem.read_ptr_sized(
-                    ref_ptr.offset(mem.pointer_size(), &mem.tcx.data_layout)?.to_ptr()?,
+                    ref_ptr.ptr_offset(mem.pointer_size(), &mem.tcx.data_layout)?.to_ptr()?,
                     align
-                )?.to_bytes()? as u64;
+                )?.to_bits(mem.pointer_size())? as u64;
                 Ok((ptr, len))
             }
-            Value::ByValPair(ptr, val) => {
-                let len = val.to_u128()?;
-                assert_eq!(len as u64 as u128, len);
+            Value::ScalarPair(ptr, val) => {
+                let len = val.to_bits(self.memory().pointer_size())?;
                 Ok((ptr.into(), len as u64))
             }
-            Value::ByVal(PrimVal::Undef) => err!(ReadUndefBytes),
-            Value::ByVal(_) => bug!("expected ptr and length, got {:?}", value),
+            Value::Scalar(Scalar::Bits { defined: 0, .. }) => err!(ReadUndefBytes),
+            Value::Scalar(_) => bug!("expected ptr and length, got {:?}", value),
```

Comment:
How is it possible to get a `Value::Scalar(Scalar::undef())` for a type with an `Abi::ScalarPair`?

Reply:
by reading a fat pointer from undefined bytes I'd presume

Reply:
That should produce a `ScalarPair` with two undefs.